File: Dot examples/ASP Book Examples/Book Examples/Chapter27/CS/App_Code/LookupExpressionBuilder.cs
using System;
using System.CodeDom;
using System.Web.UI;
using System.ComponentModel;
using System.Web.Compilation;
using System.Xml;
using System.Web.Hosting;
using System.Web.Caching;

namespace AspNetUnleashed
{
    public class LookupExpressionBuilder : ExpressionBuilder
    {
        public override CodeExpression GetCodeExpression(BoundPropertyEntry entry, object parsedData, ExpressionBuilderContext context)
        {
            CodeTypeReferenceExpression refMe = new CodeTypeReferenceExpression(base.GetType());
            CodePrimitiveExpression expression = new CodePrimitiveExpression(entry.Expression);
            return new CodeMethodInvokeExpression(refMe, "GetEvalData", new CodeExpression[] { expression });
        }

        public override object EvaluateExpression(object target, BoundPropertyEntry entry, object parsedData, ExpressionBuilderContext context)
        {
            return GetEvalData(entry.Expression);
        }

        public override bool SupportsEvaluate
        {
            get
            {
                return true;
            }
        }

        public static string GetEvalData(string expression)
        {
            XmlDocument lookupDoc = (XmlDocument)HostingEnvironment.Cache["Lookup"];
            if (lookupDoc == null)
            {
                lookupDoc = new XmlDocument();
                string lookupFileName = HostingEnvironment.MapPath("~/Lookup.config");
                lookupDoc.Load(lookupFileName);
                CacheDependency fileDepend = new CacheDependency(lookupFileName);
                HostingEnvironment.Cache.Insert("Lookup", lookupDoc, fileDepend);
            }

            string search = String.Format("//add[@key='{0}']", expression);
            XmlNode match = lookupDoc.SelectSingleNode(search);
            if (match != null)
                return match.Attributes["value"].Value;
            return "[no match]";
        }

    }
}

File: Dot examples/ASP Book Examples/Book Examples/Chapter27/VB/App_Code/LookupExpressionBuilder.vb
Imports System
Imports System.CodeDom
Imports System.Web.UI
Imports System.ComponentModel
Imports System.Web.Compilation
Imports System.Xml
Imports System.Web.Hosting
Imports System.Web.Caching

Namespace AspNetUnleashed
    Public Class LookupExpressionBuilder
        Inherits ExpressionBuilder

        Public Overrides Function GetCodeExpression(ByVal enTry As BoundPropertyEntry, ByVal parsedData As Object, ByVal context As ExpressionBuilderContext) As CodeExpression
            Dim refMe As New CodeTypeReferenceExpression(MyBase.GetType())
            Dim expression As New CodePrimitiveExpression(enTry.Expression)
            Return New CodeMethodInvokeExpression(refMe, "GetEvalData", New CodeExpression() {expression})
        End Function

        Public Overrides Function EvaluateExpression(ByVal target As Object, ByVal enTry As BoundPropertyEnTry, ByVal parsedData As Object, ByVal context As ExpressionBuilderContext) As Object
            Return GetEvalData(enTry.Expression)
        End Function

        Public Overrides ReadOnly Property SupportsEvaluate() As Boolean
            Get
                Return True
            End Get
        End Property

        Public Shared Function GetEvalData(ByVal expression As String) As String
            Dim lookupDoc As XmlDocument = CType(HostingEnvironment.Cache("Lookup"), XmlDocument)
            If IsNothing(lookupDoc) Then
                lookupDoc = New XmlDocument()
                Dim lookupFileName As String = HostingEnvironment.MapPath("~/Lookup.config")
                lookupDoc.Load(lookupFileName)
                Dim fileDepend As New CacheDependency(lookupFileName)
                HostingEnvironment.Cache.Insert("Lookup", lookupDoc, fileDepend)
            End If

            Dim search As String = String.Format("//add[@key='{0}']", expression)
            Dim match As XmlNode = lookupDoc.SelectSingleNode(search)
            If Not IsNothing(match) Then
                Return match.Attributes("value").Value
            End If
            Return "[no match]"
        End Function

    End Class
End Namespace

